service-base/src/main/java/org/eclipse/hono/service/AbstractAdapterConfig.java

```diff
@@ -266,13 +312,30 @@ public HonoConnection downstreamConnection() {
      */
     @Qualifier(TelemetryConstants.TELEMETRY_ENDPOINT)
     @Bean
+    @ConditionalOnAmqpMessaging
```

Comment:
IMHO we do not need the conditional annotations because we only instantiate the beans directly by invoking the methods explicitly ...

Reply:
Do you want to remove Spring Boot? Otherwise, we could refactor `ProtonBasedDownstreamSender` into two classes, one implementing `TelemetrySender` and one implementing `EventSender` and then look up the beans from the ApplicationContext, like you do it with the DeviceConnectionClient. This would remove the ugly if-else block from _setCollaborators()_ and instead take advantage of conditional bean instantiation based on config properties (which are used by the conditional annotations).

Reply:
Another option for the conditional bean creation would be to use _profiles_. This is probably the more common way to handle it with Spring Boot. Quarkus has a concept of profiles as well: https://quarkus.io/guides/config#configuration-profiles

Reply:
> Otherwise, we could refactor `ProtonBasedDownstreamSender` into two classes, one implementing `TelemetrySender` and one implementing `EventSender` and then look up the beans from the ApplicationContext, like you do it with the DeviceConnectionClient.

This is what #2326 provides.

Reply:
I do not mind the if-else block in *setCollaborators* for the time being. We will most probably need to do something like that for the Quarkus adapters anyway, as it doesn't support conditional instantiation based on runtime configuration ...